File: android-app/app 2/src/main/java/com/bandite/sonicwalkscape/utils/SubtitleParser.kt
package com.bandite.sonicwalkscape.utils

/**
 * Parser for SRT (SubRip) subtitle files
 */
object SubtitleParser {

    data class Subtitle(
        val index: Int,
        val startTimeMs: Long,
        val endTimeMs: Long,
        val text: String
    )

    /**
     * Parse SRT content into a list of Subtitle objects
     */
    fun parse(srtContent: String): List<Subtitle> {
        val subtitles = mutableListOf<Subtitle>()
        val blocks = srtContent.trim().split(Regex("\n\n+"))

        for (block in blocks) {
            val lines = block.trim().split("\n")
            if (lines.size < 3) continue

            try {
                val index = lines[0].trim().toIntOrNull() ?: continue
                val timeLine = lines[1]
                val text = lines.drop(2).joinToString("\n")

                val times = parseTimeLine(timeLine) ?: continue

                subtitles.add(
                    Subtitle(
                        index = index,
                        startTimeMs = times.first,
                        endTimeMs = times.second,
                        text = text.trim()
                    )
                )
            } catch (e: Exception) {
                DebugLogger.warning("Failed to parse subtitle block: ${e.message}")
            }
        }

        return subtitles.sortedBy { it.startTimeMs }
    }

    /**
     * Parse time line in format "00:00:00,000 --> 00:00:00,000"
     */
    private fun parseTimeLine(line: String): Pair<Long, Long>? {
        val parts = line.split("-->")
        if (parts.size != 2) return null

        val startTime = parseTime(parts[0].trim()) ?: return null
        val endTime = parseTime(parts[1].trim()) ?: return null

        return Pair(startTime, endTime)
    }

    /**
     * Parse time string in format "00:00:00,000" to milliseconds
     */
    private fun parseTime(timeString: String): Long? {
        // Handle both "00:00:00,000" and "00:00:00.000" formats
        val normalized = timeString.replace(",", ".")
        val parts = normalized.split(":")

        if (parts.size != 3) return null

        return try {
            val hours = parts[0].toLong()
            val minutes = parts[1].toLong()
            val secondsParts = parts[2].split(".")
            val seconds = secondsParts[0].toLong()
            val millis = if (secondsParts.size > 1) {
                secondsParts[1].padEnd(3, '0').take(3).toLong()
            } else 0L

            (hours * 3600000) + (minutes * 60000) + (seconds * 1000) + millis
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Get the subtitle text for a given time position
     */
    fun getSubtitleAt(subtitles: List<Subtitle>, timeMs: Long): String? {
        return subtitles.find { subtitle ->
            timeMs >= subtitle.startTimeMs && timeMs <= subtitle.endTimeMs
        }?.text
    }
}

File: mobile-app/ios/SonicWalkscape/SonicWalkscape/Utilities/SubtitleParser.swift
//
//  SubtitleParser.swift
//  SonicWalkscape
//
//  Parses .srt subtitle files and matches text to audio playback time
//

import Foundation

/// Represents a single subtitle entry with time codes and text
struct Subtitle {
    let startTime: TimeInterval
    let endTime: TimeInterval
    let text: String

    /// Check if this subtitle should be displayed at the given time
    func isActive(at time: TimeInterval) -> Bool {
        return time >= startTime && time <= endTime
    }
}

/// Parser for .srt (SubRip) subtitle files
class SubtitleParser {

    /// Parse .srt file content and return array of subtitles
    static func parse(_ srtContent: String) -> [Subtitle] {
        var subtitles: [Subtitle] = []

        // Split into blocks (separated by double newlines)
        let blocks = srtContent.components(separatedBy: "\n\n")

        for block in blocks {
            let lines = block.components(separatedBy: "\n").filter { !$0.isEmpty }

            // Each block should have at least: index, timecode, text
            guard lines.count >= 3 else { continue }

            // Parse time code (line 2): "00:00:05,500 --> 00:00:10,000"
            let timecodeLine = lines[1]
            guard let (startTime, endTime) = parseTimecode(timecodeLine) else {
                continue
            }

            // Text is everything from line 3 onwards
            let text = lines[2...].joined(separator: "\n")

            subtitles.append(Subtitle(
                startTime: startTime,
                endTime: endTime,
                text: text
            ))
        }

        return subtitles
    }

    /// Parse timecode string: "00:00:05,500 --> 00:00:10,000"
    private static func parseTimecode(_ line: String) -> (TimeInterval, TimeInterval)? {
        let parts = line.components(separatedBy: " --> ")
        guard parts.count == 2 else { return nil }

        guard let start = parseTime(parts[0]),
              let end = parseTime(parts[1]) else {
            return nil
        }

        return (start, end)
    }

    /// Parse time string: "00:00:05,500" to TimeInterval
    private static func parseTime(_ timeString: String) -> TimeInterval? {
        // Format: HH:MM:SS,mmm
        let cleaned = timeString.trimmingCharacters(in: .whitespaces)
        let parts = cleaned.replacingOccurrences(of: ",", with: ".").components(separatedBy: ":")

        guard parts.count == 3 else { return nil }

        guard let hours = Double(parts[0]),
              let minutes = Double(parts[1]),
              let seconds = Double(parts[2]) else {
            return nil
        }

        return hours * 3600 + minutes * 60 + seconds
    }

    /// Get the subtitle that should be displayed at a given time
    static func subtitle(at time: TimeInterval, in subtitles: [Subtitle]) -> Subtitle? {
        return subtitles.first { $0.isActive(at: time) }
    }
}
